altair/utils/data.py

```diff
@@ -69,12 +100,19 @@ class MaxRowsError(Exception):
     pass
 
 
-@curried.curry
-def limit_rows(data: TDataType, max_rows: Optional[int] = 5000) -> TDataType:
+@overload
+def limit_rows(data: Literal[None] = ..., max_rows: Optional[int] = ...) -> partial: ...
```

Comment:
What's the purpose of setting the default parameter values to elipsis here (`Literal[None] = ...`)?

Reply:
There's two parts to this.

# Replicating the existing behaviour provided by `@curried.curry`
This isn't immediately obvious when seeing the original signature as: 
```py
@curried.curry
def limit_rows(data: TDataType, max_rows: Optional[int] = 5000) -> TDataType:...
```
but the [curry](https://toolz.readthedocs.io/en/latest/curry.html#id1) [decorator](https://toolz.readthedocs.io/en/latest/_modules/toolz/functoolz.html#curry) roughly* alters it to:
```py
def limit_rows(data: Optional[TDataType] = None, max_rows: Optional[int] = 5000) -> TDataType:...
```
*with the caveats that:
- `TDataType` is only returned once all non-optional arguments have been provided
- Until that condition is met, other arguments can be provided again and alter the function

With this in mind, the correctly typed definition has this default `None` already there - as this was implicit before and hidden behind the decorator.

# Representing default arguments within `@typing.overload`
`...` in an `@overload` will match without the user providing an argument. Otherwise, they would need to write the below for the a match:
```py
limit_rows(max_rows=20) # fails
limit_rows(None, 5000) # matches
```
With the current `@overload` defintion, both are valid and match the same `@overload` def.

## Side note
While I was searching through the [typing spec](https://typing.readthedocs.io/en/latest/source/stubs.html#functions-and-methods) for use of `...` I came across [Literal[None] conversion](https://typing.readthedocs.io/en/latest/spec/literal.html#legal-parameters-for-literal-at-type-check-time). 
Which was news to me, so I'm going to simplify that part - and some places in #3431.

Reply:
Thanks for elaborating, I think I understand now... something along the lines of that `...` is a "typeless" default argument that is useful where `None` would not satisfy the type constraint, such as for `overload` decorators.

The simplification for `Literal[None]` is welcome, good find.

Reply:
@joelostblom Yeah you're pretty much there. 

If you're familiar with [pandas.DataFrame.rename](https://github.com/pandas-dev/pandas/blob/de5d7323cf6fcdd6fcb1643a11c248440787d960/pandas/core/frame.py#L5440) you could try that out in your IDE to see how different `overload`'s get matched. 

I picked that one as it has arguments without `None` - as that part of it isn't directly linked to `...` - it is more about matching the already defined default. 

So for `errors` the `...` represents *'ignore'* by default, or either *'ignore'*, *'raise'* if provided explicitly:
```py
class DataFrame:
    @overload
	def rename(
        self,
        errors: Literal['ignore', 'raise'] = ...,
    ) -> DataFrame | None:
        ...

    def rename(
        self,
        errors: Literal['ignore', 'raise'] = "ignore",
    ) -> DataFrame | None:
```

Reply:
> The simplification for `Literal[None]` is welcome, good find.

Added in: a8e537f1547a9547e1dfc75e6848574829fcdd8a